## Q1. A company conducted a survey of its employees and found that 70% of the employees use the company's health insurance plan, while 40% of the employees who use the plan are smokers. What is the probability that an employee is a smoker given that he/she uses the health insurance plan?

The probability that an employee is a smoker given that he/she uses the health insurance plan is 40%.

Code snippet
P(Smoker|Uses Plan) = P(Smoker and Uses Plan) / P(Uses Plan)
= 0.4 / 0.7
= 0.5714285714285714
Use code with caution. Learn more
This means that if you randomly select an employee who uses the health insurance plan, there is a 57.14% chance that they are a smoker.

Here is a breakdown of the probabilities:

P(Smoker) = 0.30 (30%)
P(Uses Plan) = 0.70 (70%)
P(Smoker and Uses Plan) = 0.40 (40%)
The probability that an employee is a smoker and uses the health insurance plan is equal to the probability that an employee is a smoker multiplied by the probability that an employee uses the health insurance plan

P(Smoker and Uses Plan) = P(Smoker) * P(Uses Plan)
= 0.30 * 0.70
= 0.21


The probability that an employee is a smoker given that he/she uses the health insurance plan is 40%.

Code snippet
P(Smoker|Uses Plan) = P(Smoker and Uses Plan) / P(Uses Plan)
= 0.4 / 0.7
= 0.5714285714285714
Use code with caution. Learn more
This means that if you randomly select an employee who uses the health insurance plan, there is a 57.14% chance that they are a smoker.

Here is a breakdown of the probabilities:

P(Smoker) = 0.30 (30%)
P(Uses Plan) = 0.70 (70%)
P(Smoker and Uses Plan) = 0.40 (40%)
The probability that an employee is a smoker and uses the health insurance plan is equal to the probability that an employee is a smoker multiplied by the probability that an employee uses the health insurance plan.

Code snippet
P(Smoker and Uses Plan) = P(Smoker) * P(Uses Plan)
= 0.30 * 0.70
= 0.21

The probability that an employee uses the health insurance plan is given by the survey results.

P(Uses Plan) = 0.70

The probability that an employee is a smoker given that he/she uses the health insurance plan is calculated by dividing the probability that an employee is a smoker and uses the health insurance plan by the probability that an employee uses the health insurance plan.

P(Smoker|Uses Plan) = P(Smoker and Uses Plan) / P(Uses Plan)
= 0.21 / 0.70
= 0.5714285714285714


## Q2. What is the difference between Bernoulli Naive Bayes and Multinomial Naive Bayes?

Bernoulli Naive Bayes and Multinomial Naive Bayes are two variants of the Naive Bayes classifier that are commonly used for different types of data. Here are the key differences between the two:

`Feature Type:`

- Bernoulli Naive Bayes: It is suitable for binary or Boolean features, where each feature can take only two values, typically representing the presence or absence of a particular attribute.
- Multinomial Naive Bayes: It is designed for discrete features that can have multiple values or represent counts or frequencies. It is commonly used in text classification tasks where features are typically word occurrences or frequencies.

`Feature Distribution:`

- Bernoulli Naive Bayes: It assumes that the features are independent binary variables, and their distribution is modeled using a Bernoulli distribution. Each feature is considered as a binary variable, representing its presence or absence in a particular instance.
- Multinomial Naive Bayes: It assumes that the features follow a multinomial distribution. The features are typically treated as counts or frequencies of events in a document or a sample, such as word occurrences or term frequencies.

`Handling Absence of Features:`

- Bernoulli Naive Bayes: It explicitly models the absence of features and considers it as an informative factor. The absence of a feature contributes to the classification decision.
- Multinomial Naive Bayes: It does not explicitly model the absence of features. It focuses on the presence and frequency of features, assuming that the absence of a feature does not provide additional information. 

`Data Representation:`

- Bernoulli Naive Bayes: It often uses a binary feature vector representation, where each feature is represented as a binary value (0 or 1), indicating its presence or absence.
- Multinomial Naive Bayes: It typically uses a frequency-based representation, where each feature is represented by its count or frequency in a document or sample.

In summary, Bernoulli Naive Bayes is suitable for binary features, explicitly models the absence of features, and follows a Bernoulli distribution. Multinomial Naive Bayes, on the other hand, is used for discrete features or word frequencies, does not explicitly model the absence of features, and assumes a multinomial distribution. The choice between the two depends on the nature of the features and the specific requirements of the classification problem.

## Q3. How does Bernoulli Naive Bayes handle missing values?

Bernoulli Naive Bayes handles missing values by treating them as a separate category or value for each feature. Instead of assuming the absence or presence of a feature, a missing value is considered as a distinct category or value within the feature.

When training a Bernoulli Naive Bayes classifier, the presence or absence of each feature is determined based on the observed data. However, if a missing value is encountered during prediction for a particular feature, the classifier considers it as a separate category and calculates probabilities accordingly.

Here's a step-by-step explanation of how Bernoulli Naive Bayes handles missing values:

`During training:`

- For each feature, the presence or absence of the feature is recorded based on the observed data. If a feature is present in a training instance, its corresponding value is set to 1; otherwise, it is set to 0.

`During prediction:`

- If a missing value is encountered for a particular feature in the test instance, the classifier treats it as a separate category or value within that feature.

- The probability calculations for each class consider the missing value as a distinct category and include it in the calculations.


By treating missing values as a separate category, Bernoulli Naive Bayes allows for the inclusion of instances with missing values during the classification process. However, it's important to note that the treatment of missing values in Bernoulli Naive Bayes depends on the specific implementation and may vary in different scenarios.

## Q4. Can Gaussian Naive Bayes be used for multi-class classification?

Yes, Gaussian Naive Bayes can be used for multi-class classification. Gaussian Naive Bayes is a variant of the Naive Bayes classifier that assumes the features within each class are normally distributed (follow a Gaussian distribution). It is commonly used for continuous numerical features.

To adapt Gaussian Naive Bayes for multi-class classification, the classifier uses the same underlying principles as in binary classification but extends them to handle multiple classes. The classifier calculates the likelihoods and prior probabilities for each class and then applies Bayes' theorem to calculate the posterior probabilities.

The steps for using Gaussian Naive Bayes for multi-class classification are as follows:

`Training:`
- For each class, calculate the mean and standard deviation of each feature based on the training data belonging to that class.

- Estimate the prior probability for each class by calculating the frequency or proportion of training instances belonging to each class.

`Prediction:`

- Given a new instance with feature values, calculate the likelihoods for each class using the Gaussian probability density function (PDF) based on the mean and standard deviation of each feature in each class.
- Multiply the likelihoods by the prior probabilities for each class.
- Normalize the probabilities by dividing each class probability by the sum of all class probabilities to obtain posterior probabilities.
- The class with the highest posterior probability is predicted as the class label for the new instance.
- In multi-class classification with Gaussian Naive Bayes, the model assumes that each feature follows a Gaussian distribution within each class, and the class probabilities are calculated based on these distributions. It makes class predictions based on the posterior probabilities.

It's important to note that Gaussian Naive Bayes assumes that the features are conditionally independent given the class label. This is a simplifying assumption and may not hold in all cases. However, despite this assumption, Gaussian Naive Bayes can still perform well in practice, especially when the feature independence assumption is reasonably satisfied or when there are limited training data available.

## Q5. Assignment:
### `Data preparation`:
- Download the "Spambase Data Set" from the UCI Machine Learning Repository (https://archive.ics.uci.edu/ml/datasets/Spambase). This dataset contains email messages, where the goal is to predict whether a message is spam or not based on several input features.

In [13]:
import pandas as pd

# Define the feature names
feature_names = [
    'word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d', 'word_freq_our',
    'word_freq_over', 'word_freq_remove', 'word_freq_internet', 'word_freq_order', 'word_freq_mail',
    'word_freq_receive', 'word_freq_will', 'word_freq_people', 'word_freq_report', 'word_freq_addresses',
    'word_freq_free', 'word_freq_business', 'word_freq_email', 'word_freq_you', 'word_freq_credit',
    'word_freq_your', 'word_freq_font', 'word_freq_000', 'word_freq_money', 'word_freq_hp', 'word_freq_hpl',
    'word_freq_george', 'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet', 'word_freq_857',
    'word_freq_data', 'word_freq_415', 'word_freq_85', 'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
    'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting', 'word_freq_original', 'word_freq_project',
    'word_freq_re', 'word_freq_edu', 'word_freq_table', 'word_freq_conference', 'char_freq_;', 'char_freq_(', 'char_freq_[',
    'char_freq_!', 'char_freq_$', 'char_freq_#', 'capital_run_length_average', 'capital_run_length_longest',
    'capital_run_length_total', 'spam'
]

# Read the spambase.data file into a pandas DataFrame
data = pd.read_csv('spambase.data', header=None)

# Assign the feature names as column headers
data.columns = feature_names

# Print the DataFrame
data.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

### `Implementation:`
Implement Bernoulli Naive Bayes, Multinomial Naive Bayes, and Gaussian Naive Bayes classifiers using the
scikit-learn library in Python. Use 10-fold cross-validation to evaluate the performance of each classifier on the
dataset. You should use the default hyperparameters for each classifier.

In [14]:
import pandas as pd
from sklearn.model_selection import cross_val_score
from sklearn.naive_bayes import BernoulliNB, MultinomialNB, GaussianNB

# Read the spambase.data file into a pandas DataFrame
data = pd.read_csv('spambase.data', header=None)

# Assign the feature names as column headers
data.columns = feature_names

# Extract the features and target variable
X = data.drop('spam', axis=1)
y = data['spam']

# Initialize the classifiers
bernoulli_nb = BernoulliNB()
multinomial_nb = MultinomialNB()
gaussian_nb = GaussianNB()

# Perform 10-fold cross-validation and evaluate the performance
print("Bernoulli Naive Bayes:")
bernoulli_scores = cross_val_score(bernoulli_nb, X, y, cv=10)
print("Accuracy:", bernoulli_scores.mean())

print("\nMultinomial Naive Bayes:")
multinomial_scores = cross_val_score(multinomial_nb, X, y, cv=10)
print("Accuracy:", multinomial_scores.mean())

print("\nGaussian Naive Bayes:")
gaussian_scores = cross_val_score(gaussian_nb, X, y, cv=10)
print("Accuracy:", gaussian_scores.mean())


Bernoulli Naive Bayes:
Accuracy: 0.8839380364047911

Multinomial Naive Bayes:
Accuracy: 0.7863496180326323

Gaussian Naive Bayes:
Accuracy: 0.8217730830896915



### `Results:`
Report the following performance metrics for each classifier:
- Accuracy
- Precision
- Recall
- F1 score

In [16]:

# Initialize the classifiers
bernoulli_nb = BernoulliNB()
multinomial_nb = MultinomialNB()
gaussian_nb = GaussianNB()

# Perform 10-fold cross-validation and evaluate the performance
print("Bernoulli Naive Bayes:")
bernoulli_accuracy = cross_val_score(bernoulli_nb, X, y, cv=10, scoring='accuracy').mean()
bernoulli_precision = cross_val_score(bernoulli_nb, X, y, cv=10, scoring='precision').mean()
bernoulli_recall = cross_val_score(bernoulli_nb, X, y, cv=10, scoring='recall').mean()
bernoulli_f1 = cross_val_score(bernoulli_nb, X, y, cv=10, scoring='f1').mean()

print("Accuracy:", bernoulli_accuracy)
print("Precision:", bernoulli_precision)
print("Recall:", bernoulli_recall)
print("F1 Score:", bernoulli_f1)

print("\nMultinomial Naive Bayes:")
multinomial_accuracy = cross_val_score(multinomial_nb, X, y, cv=10, scoring='accuracy').mean()
multinomial_precision = cross_val_score(multinomial_nb, X, y, cv=10, scoring='precision').mean()
multinomial_recall = cross_val_score(multinomial_nb, X, y, cv=10, scoring='recall').mean()
multinomial_f1 = cross_val_score(multinomial_nb, X, y, cv=10, scoring='f1').mean()

print("Accuracy:", multinomial_accuracy)
print("Precision:", multinomial_precision)
print("Recall:", multinomial_recall)
print("F1 Score:", multinomial_f1)

print("\nGaussian Naive Bayes:")
gaussian_accuracy = cross_val_score(gaussian_nb, X, y, cv=10, scoring='accuracy').mean()
gaussian_precision = cross_val_score(gaussian_nb, X, y, cv=10, scoring='precision').mean()
gaussian_recall = cross_val_score(gaussian_nb, X, y, cv=10, scoring='recall').mean()
gaussian_f1 = cross_val_score(gaussian_nb, X, y, cv=10, scoring='f1').mean()

print("Accuracy:", gaussian_accuracy)
print("Precision:", gaussian_precision)
print("Recall:", gaussian_recall)
print("F1 Score:", gaussian_f1)

Bernoulli Naive Bayes:
Accuracy: 0.8839380364047911
Precision: 0.8869617393737383
Recall: 0.8152389047416673
F1 Score: 0.8481249015095276

Multinomial Naive Bayes:
Accuracy: 0.7863496180326323
Precision: 0.7393175533565436
Recall: 0.7214983911116508
F1 Score: 0.7282909724016348

Gaussian Naive Bayes:
Accuracy: 0.8217730830896915
Precision: 0.7103733928118492
Recall: 0.9569516119239877
F1 Score: 0.8130660909542995


### `Discussion:`
Discuss the results you obtained. Which variant of Naive Bayes performed the best? Why do you think that is
the case? Are there any limitations of Naive Bayes that you observed?

Based on the accuracy metric alone, Bernoulli Naive Bayes performed the best with an accuracy of 0.8839. It achieved the highest accuracy among the three classifiers. However, it's important to note that accuracy alone may not provide a complete picture of the classifier's performance.

The reason Bernoulli Naive Bayes performed well could be due to the nature of the dataset and the assumption of binary features in the Bernoulli model. Since the dataset contains binary features (presence or absence of words), the Bernoulli Naive Bayes model, which assumes binary features, may be a better fit for this specific dataset.

On the other hand, Multinomial Naive Bayes, which is commonly used for text classification with discrete features, achieved lower accuracy. This could be because the dataset contains continuous features (e.g., word frequencies) that are not well-suited for the Multinomial Naive Bayes assumption.

Gaussian Naive Bayes, which assumes a Gaussian distribution for continuous features, performed reasonably well with an accuracy of 0.8218. However, its precision is lower compared to the other two classifiers. This could be because the dataset may not strictly follow a Gaussian distribution for all features.

Limitations of Naive Bayes:

Naive Bayes assumes independence between features, which may not hold true in real-world scenarios. This assumption can affect the performance of the classifier.
It may struggle with rare feature combinations or unseen data during training.
Naive Bayes can be sensitive to irrelevant features or features that are not informative for the target variable.
It assumes that all features contribute equally to the classification, which may not be the case in reality.
These limitations should be taken into consideration when applying Naive Bayes classifiers and evaluating their performance. It's always important to assess the suitability of the algorithm for the specific dataset and problem at hand.


### `Conclusion:`
Summarise your findings and provide some suggestions for future work.

Suggestions for future work:

Feature engineering: Explore different feature engineering techniques to enhance the performance of the classifiers. For example, consider transforming continuous features into binary features or applying different scaling methods to improve the modeling assumptions of the classifiers.

Hyperparameter tuning: Perform hyperparameter tuning for each classifier to find optimal parameter settings that can potentially improve their performance.

Ensemble methods: Investigate the use of ensemble methods, such as combining multiple Naive Bayes classifiers or using ensemble techniques like bagging or boosting, to further enhance the predictive power of the models.

Evaluate other classifiers: Compare the performance of Naive Bayes with other classification algorithms, such as decision trees, support vector machines, or neural networks, to identify the best-performing model for the spambase dataset.

Handling class imbalance: If the dataset exhibits class imbalance, apply techniques such as oversampling or undersampling to address the imbalance and improve the performance of the classifiers.

Cross-validation strategies: Experiment with different cross-validation strategies, such as stratified cross-validation or leave-one-out cross-validation, to assess the robustness of the models and obtain more reliable performance estimates.

Feature selection: Explore feature selection techniques, such as information gain, chi-square test, or recursive feature elimination, to identify the most informative features and reduce the dimensionality of the dataset.

By considering these suggestions and further exploring the limitations and potential improvements of Naive Bayes classifiers, future work can lead to enhanced spam email classification performance and potentially contribute to the development of more effective anti-spam systems.